File: saucelabs-selenium-project/pages/products_page.py
"""
products_page.py - Page Object for the SauceDemo inventory/products page.
URL: https://www.saucedemo.com/inventory.html
"""
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import Select
from pages.base_page import BasePage


class ProductsPage(BasePage):
    """Page Object for the Products listing page."""

    # Locators
    PAGE_TITLE = (By.CLASS_NAME, "title")
    PRODUCT_ITEMS = (By.CLASS_NAME, "inventory_item")
    PRODUCT_NAMES = (By.CLASS_NAME, "inventory_item_name")
    PRODUCT_DESCRIPTIONS = (By.CLASS_NAME, "inventory_item_desc")
    PRODUCT_PRICES = (By.CLASS_NAME, "inventory_item_price")
    PRODUCT_IMAGES = (By.CSS_SELECTOR, ".inventory_item_img img")
    SORT_DROPDOWN = (By.CLASS_NAME, "product_sort_container")
    ADD_TO_CART_BUTTONS = (By.CSS_SELECTOR, "button[id^='add-to-cart']")
    REMOVE_BUTTONS = (By.CSS_SELECTOR, "button[id^='remove']")

    def get_page_header(self):
        return self.get_text(self.PAGE_TITLE)

    def get_all_product_names(self):
        elements = self.wait_for_all_visible(self.PRODUCT_NAMES)
        return [el.text for el in elements]

    def get_all_product_prices(self):
        elements = self.wait_for_all_visible(self.PRODUCT_PRICES)
        return [el.text for el in elements]

    def get_product_count(self):
        return len(self.driver.find_elements(*self.PRODUCT_ITEMS))

    def sort_by_name_az(self):
        Select(self.driver.find_element(*self.SORT_DROPDOWN)).select_by_value("az")
        return self

    def sort_by_name_za(self):
        Select(self.driver.find_element(*self.SORT_DROPDOWN)).select_by_value("za")
        return self

    def sort_by_price_low_high(self):
        Select(self.driver.find_element(*self.SORT_DROPDOWN)).select_by_value("lohi")
        return self

    def sort_by_price_high_low(self):
        Select(self.driver.find_element(*self.SORT_DROPDOWN)).select_by_value("hilo")
        return self

    def add_product_by_name(self, product_name):
        """Add a product to cart by its display name."""
        button_id = "add-to-cart-" + product_name.lower().replace(" ", "-")
        self.click((By.ID, button_id))
        return self

    def remove_product_by_name(self, product_name):
        button_id = "remove-" + product_name.lower().replace(" ", "-")
        self.click((By.ID, button_id))
        return self

    def click_product_by_name(self, product_name):
        """Navigate to product detail page."""
        locator = (By.XPATH, f"//div[text()='{product_name}']")
        self.click(locator)

    def is_products_page_displayed(self):
        return self.is_displayed(self.PAGE_TITLE) and self.get_page_header() == "Products"

    def are_products_listed(self):
        return self.get_product_count() > 0

    def are_all_images_displayed(self):
        images = self.driver.find_elements(*self.PRODUCT_IMAGES)
        return all(img.is_displayed() for img in images)

    def is_sort_dropdown_displayed(self):
        return self.is_displayed(self.SORT_DROPDOWN)

File: src/main/java/com/saucelabs/pages/ProductsPage.java
package com.saucelabs.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.Select;

import java.util.List;
import java.util.stream.Collectors;

/**
 * ProductsPage - Page Object for the SauceDemo inventory/products page.
 * URL: https://www.saucedemo.com/inventory.html
 * Locators are defined as By constants; no @FindBy / PageFactory used.
 */
public class ProductsPage extends BasePage {

    // Locators
    private static final By PAGE_TITLE      = By.className("title");
    private static final By PRODUCT_ITEMS   = By.className("inventory_item");
    private static final By PRODUCT_NAMES   = By.className("inventory_item_name");
    private static final By PRODUCT_PRICES  = By.className("inventory_item_price");
    private static final By PRODUCT_IMAGES  = By.cssSelector(".inventory_item_img img");
    private static final By SORT_DROPDOWN   = By.className("product_sort_container");

    public ProductsPage(WebDriver driver) {
        super(driver);
    }

    public String getPageHeader() {
        return getText(PAGE_TITLE);
    }

    public List<String> getAllProductNames() {
        return driver.findElements(PRODUCT_NAMES).stream()
                .map(WebElement::getText)
                .collect(Collectors.toList());
    }

    public List<String> getAllProductPrices() {
        return driver.findElements(PRODUCT_PRICES).stream()
                .map(WebElement::getText)
                .collect(Collectors.toList());
    }

    public int getProductCount() {
        return driver.findElements(PRODUCT_ITEMS).size();
    }

    public ProductsPage sortByNameAZ() {
        new Select(driver.findElement(SORT_DROPDOWN)).selectByValue("az");
        return this;
    }

    public ProductsPage sortByNameZA() {
        new Select(driver.findElement(SORT_DROPDOWN)).selectByValue("za");
        return this;
    }

    public ProductsPage sortByPriceLowHigh() {
        new Select(driver.findElement(SORT_DROPDOWN)).selectByValue("lohi");
        return this;
    }

    public ProductsPage sortByPriceHighLow() {
        new Select(driver.findElement(SORT_DROPDOWN)).selectByValue("hilo");
        return this;
    }

    /**
     * Add a product to cart by its URL-safe name (e.g. "sauce-labs-backpack").
     */
    public ProductsPage addProductByName(String productName) {
        String buttonId = "add-to-cart-" + productName.toLowerCase().replace(" ", "-");
        click(By.id(buttonId));
        return this;
    }

    public ProductsPage removeProductByName(String productName) {
        String buttonId = "remove-" + productName.toLowerCase().replace(" ", "-");
        click(By.id(buttonId));
        return this;
    }

    /**
     * Click a product name to navigate to its detail page.
     */
    public void clickProductByName(String productName) {
        By locator = By.xpath("//div[text()='" + productName + "']");
        click(locator);
    }

    public boolean isProductsPageDisplayed() {
        return isDisplayed(PAGE_TITLE) && "Products".equals(getPageHeader());
    }

    public boolean areProductsListed() {
        return getProductCount() > 0;
    }

    public boolean areAllImagesDisplayed() {
        List<WebElement> images = driver.findElements(PRODUCT_IMAGES);
        if (images.isEmpty()) {
            return false;
        }
        waitForVisible(images.get(0));
        return images.stream().allMatch(WebElement::isDisplayed);
    }

    public boolean isSortDropdownDisplayed() {
        return isDisplayed(SORT_DROPDOWN);
    }
}
